diff -r -u fltk-1.3.3-orig/src/fl_line_style.cxx fltk-1.3.3/src/fl_line_style.cxx
--- fltk-1.3.3-orig/src/fl_line_style.cxx	2012-03-18 19:48:29.000000000 +0100
+++ fltk-1.3.3/src/fl_line_style.cxx	2016-05-05 19:35:55.000000000 +0200
@@ -35,8 +35,8 @@
 
 #ifdef __APPLE_QUARTZ__
 float fl_quartz_line_width_ = 1.0f;
-static enum CGLineCap fl_quartz_line_cap_ = kCGLineCapButt;
-static enum CGLineJoin fl_quartz_line_join_ = kCGLineJoinMiter;
+static CGLineCap fl_quartz_line_cap_ = kCGLineCapButt;
+static CGLineJoin fl_quartz_line_join_ = kCGLineJoinMiter;
 static CGFloat *fl_quartz_line_pattern = 0;
 static int fl_quartz_line_pattern_size = 0;
 void fl_quartz_restore_line_style_() {
@@ -110,9 +110,9 @@
   DeleteObject(fl_current_xmap->pen);
   fl_current_xmap->pen = newpen;
 #elif defined(__APPLE_QUARTZ__)
-  static enum CGLineCap Cap[4] = { kCGLineCapButt, kCGLineCapButt, 
+  static CGLineCap Cap[4] = { kCGLineCapButt, kCGLineCapButt,
                                    kCGLineCapRound, kCGLineCapSquare };
-  static enum CGLineJoin Join[4] = { kCGLineJoinMiter, kCGLineJoinMiter, 
+  static CGLineJoin Join[4] = { kCGLineJoinMiter, kCGLineJoinMiter,
                                     kCGLineJoinRound, kCGLineJoinBevel };
   if (width<1) width = 1;
   fl_quartz_line_width_ = (float)width; 
